lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -1416,53 +1416,111 @@ static EnableOp getLastEnableOp(SeqOp parent) {
-///     seq { ... calyx.enable @B }           seq { ... }
-///   }                                     }
-///                                         calyx.enable @B
-static LogicalResult eliminateCommonTailEnable(IfOp ifOp,
-                                               PatternRewriter &rewriter) {
-  // Check if the branches exist.
+/// Returns a mapping of {Enabled Group, EnableOp} for all EnableOps within the
+/// immediate ParOp's body.
+static llvm::StringMap<EnableOp> getAllEnableOpsInImmediateBody(ParOp parent) {
+  llvm::StringMap<EnableOp> enables;
+  Block *body = parent.getBody();
+  for (EnableOp op : body->getOps<EnableOp>())
+    enables.insert(std::pair(op.groupName(), op));
+
+  return enables;
+}
+
+/// Removes common tail enable operations for then and else regions encompassed
+/// in SeqOp and ParOp. This canonicalization is stringent about not entering
+/// nested control operations, as this may cause unintentional changes in
+/// behavior.
+static LogicalResult eliminateCommonTail(IfOp ifOp, PatternRewriter &rewriter) {
   if (!ifOp.thenRegionExists() || !ifOp.elseRegionExists())
     return failure();
+  rewriter.setInsertionPointAfter(ifOp);
 
-  auto &thenOpStructureOp = ifOp.getThenBody()->front();
-  auto &elseOpStructureOp = ifOp.getElseBody()->front();
-  // TODO(circt/#1861): ParOps have less restrictive conditions.
-  if (isa<ParOp>(thenOpStructureOp) || isa<ParOp>(elseOpStructureOp))
-    return failure();
-
-  // At this point, only sequential operations are valid inside the branches.
-  auto thenSeqOp = dyn_cast<SeqOp>(thenOpStructureOp);
-  auto elseSeqOp = dyn_cast<SeqOp>(elseOpStructureOp);
-  assert(thenSeqOp && elseSeqOp &&
-         "expected nested seq ops in both branches of a calyx.IfOp");
-
-  EnableOp lastThenEnableOp = getLastEnableOp(thenSeqOp);
-  EnableOp lastElseEnableOp = getLastEnableOp(elseSeqOp);
+  auto &thenControl = ifOp.getThenBody()->front();
+  auto &elseControl = ifOp.getElseBody()->front();
+  if (auto r1 = dyn_cast<ParOp>(thenControl);
+      auto r2 = dyn_cast<ParOp>(elseControl)) {
+    ///   if %a with @G {                  if %a with @G {
+    ///     par {                            par { ... }
+    ///       ...                          } else {
+    ///       calyx.enable @A     ->         par { ... }
+    ///       calyx.enable @B              }
+    ///     }                              calyx.enable @A
+    ///   else {                           calyx.enable @B
+    ///     par {
+    ///       ...
+    ///       calyx.enable @A
+    ///       calyx.enable @B
+    ///     }
+    ///   }
+    llvm::StringMap<EnableOp> A = getAllEnableOpsInImmediateBody(r1),
+                              B = getAllEnableOpsInImmediateBody(r2);
+
+    // Compute the intersection between `A` and `B`.
+    SmallVector<StringRef> groupNames;
+    for (auto a = A.begin(); a != A.end(); ++a) {
+      StringRef groupName = a->getValue().groupName();
```

Comment:
`A` is a mapping of `std::pair(op.groupName(), op)`. Does this not mean that `a->getKey() == a->getValue().groupName())`?

Reply:
Yeah you're probably right I'll fix this.